### PR TITLE
Use tooltip-support-web as global tooltip provider and remove unneccessary ClayTooltipProvider components

### PR DESCRIPTION
**Pull request by bryceosterhaus**
cc @matuzalemsteles @wincent @markocikos

I was taking a look at tooltips in DXP and I think we can get away with purely using the `tooltip-support-web` module instead of making changes to how the ClayTooltipProvider component works.

1. I added `[title]:not(data-tooltip-disable)` and `[data-title]:not(data-tooltip-disable)` as selectors for our global tooltip.
2. I removed any use of just `<ClayTooltipProvider>` where we didn't need any customization.
3. Some uses were only using it for the `delay` prop, so I just added that as a `data-tooltip-delay` attribute
4. For uses of a custom `contentRenderer`, I replaced those titles with the html version of that custom component by utilizing `ReactDOMServer.renderToString`.

The last part is the one I am must unsure about. Anyone have any concerns or see issues with this?

### CODE REVIEW

**Comment by markocikos** [comment]
@bryceosterhaus This is basically what I did in an early version of https://github.com/liferay-frontend/liferay-portal/pull/946. The idea for it started in https://github.com/liferay-frontend/liferay-portal/pull/946#issuecomment-811020792. There is a lengthy discussion on it after, as well as discoveries of some ill effects of the solution. I force-pushed over it later on, but it was the same in principle: have a global trigger on `title`, remove `<ClayTooltipProvider>`s. I did not do the typescript migration, :+1: , we should do that at some point regardless of this ticket.

You can find more details in https://github.com/liferay-frontend/liferay-portal/pull/946, but here is summary of problems with the solution:
- [minor] It broke some CI tests, that are done over `title` attribute.
- [minor] There was a bug with empty `title` attribute we like to place in tags.
- [minor] There was a bug with `title` attribute on an iframe. It has a different meaning, should not show a tooltip.
- [minor] There is a bug with tooltip alignment, where only default alignment works. This bug is not in `<ClayTooltipProvider>`. It does not show in current usages.
- [moderate] ClayTooltipProvider has some advanced features that we would be losing, such as custom renderer. This is used in few places.
- [blocker] A global trigger on `title` attribute spills over to customer portlets. They may have or want to have their own custom tooltips solution. support-web is very aggressive: we are removing `title` attribute on hover, to prevent default browser behavior. This suppresses everything else that relies on the attribute, including `<ClayTooltipProvider>` and potential customer solutions. The existing solution in support-web reflects the practice of being non-invasive. It is scoped specifically to few portlets, or a special `lfr-portal-tooltip` class.

I dislike the solution of just adding a list of individual portlets. I tried to scope "all Liferay-developed modules", but found this impossible. We have quite a few solutions that generate nondescript HTML appended to body (Portals, dropdowns etc.). In the end, I gave up and threw the hot potato to Matu, to see if we can have a good solution on the Clay side.

**Comment by matuzalemsteles** [comment]
Well, I think the only problem I see with this is to use `ReactDOMServer.renderToString` to simulate what we did with `contentRenderer`. Maybe for those cases, we can just leave `ClayTooltipProvider`?

I think we should also avoid creating our own Tooltip solution within DXP, because as @markocikos said there are some features that `ClayTooltipProvider` has and this implementation doesn’t, so I think we can make the tooltip-support-web use `ClayTooltipProvider` that will provide the main features and we make `ClayTooltipProvider` more configurable, like telling where to add listeners or if you need exclusion selectors. Then we will have:

- `tooltip-support-web` global based on `ClayTooltipProvider` to solve some problems that @markocikos cited
- `ClayTooltipProvider` for more advanced use cases where `contentRenderer` is used

In theory, we won't be adding by default the `ClayTooltipProvider`, the teams can use as they when need to customize things more but at least we centralize the logic on top of` ClayTooltipProvider`. I still think that we would have to tackle the problems that Marko cited independently once we made the global `ClayTooltipProvider` being used in `tooltip-support-web`

We will still have problems with `tooltip-support-web` being invasive even though we put `data-tooltip-disable` as support to disable, well this is very complicated and I think that in any direction we are going to have some loss.

**Comment by bryceosterhaus** [comment]
> [minor] It broke some CI tests, that are done over title attribute.
> [minor] There was a bug with empty title attribute we like to place in tags.
> [minor] There was a bug with title attribute on an iframe. It has a different meaning, should not show a tooltip.

I think these three should be manageable with a little bit of work, but none of these seem to be a big hurdle.

> [moderate] There is a bug with tooltip alignment, where only default alignment works. This bug is not in <ClayTooltipProvider>. It does not show in current usages.

We should be able to easily copy this logic over from ClayTooltipProvider and use the domAlign library.

> [major] ClayTooltipProvider has some advanced features that we would be losing, such as custom renderer. This is used in few places.

Custom renderer should be possible using `ReactDOMServer.renderToString()`. I have a few examples where I ported that over as well.

> [blocker] A global trigger on title attribute spills over to customer portlets. They may have or want to have their own custom tooltips solution. support-web is very aggressive: we are removing title attribute on hover, to prevent default browser behavior. This suppresses everything else that relies on the attribute, including <ClayTooltipProvider> and potential customer solutions. The existing solution in support-web reflects the practice of being non-invasive. It is scoped specifically to few portlets, or a special lfr-portal-tooltip class.

This seems is indeed a concern. Do we have any examples where customers are asking for custom implementations of Tooltip and also do we have examples where they want to leverage our Tooltip? One thought would be to have an "opt-out" class they could add to their portlet, but that means they have to make a change.

This seems like something we would have already supported well in previous DXP versions, anyone know why its become an issue for us now and wasn't back for previous 7.x versions?

**Comment by bryceosterhaus** [comment]
And thanks for the additional context, @markocikos! I must have missed your PR when I was catching up on notifications after leave. I didn't realize you already spent time on it.

**Comment by markocikos** [comment]
@bryceosterhaus 

> Do we have any examples where customers are asking for custom implementations of Tooltip and also do we have examples where they want to leverage our Tooltip? One thought would be to have an "opt-out" class they could add to their portlet, but that means they have to make a change.

I wondered about the same thing during discussion in https://github.com/liferay-frontend/liferay-portal/pull/946. What I got as a counterpoint from @jbalsas is that opt-in is better than opt-out (see https://github.com/liferay-frontend/liferay-portal/pull/946#issuecomment-819533360 and https://github.com/liferay-frontend/liferay-portal/pull/946#issuecomment-819306030). It does make sense to me, opt-in seems like a better user experience. A potential for a "wow moment" vs. a potential for frustration not knowing how to get rid of an unwanted feature. 

Of course, this is nowhere near black & white case, we're talking about layers over layers of subjective calls, both on our side and on customers' side. I don't know if there was any targeted research on customers. I doubt there were any examples, like tickets requesting capability for own solution, because we were enabling custom solutions for a long time.

>This seems like something we would have already supported well in previous DXP versions, anyone know why its become an issue for us now and wasn't back for previous 7.x versions?

This is our current opt-in mechanism: add `lfr-portal-tooltip` on element with `title`. Many old tags and solutions do this, see for example [icon tag](https://github.com/liferay/liferay-portal/blob/master/portal-web/docroot/html/taglib/ui/icon/init.jsp#L90-L92) and [input tag](https://github.com/liferay/liferay-portal/blob/master/portal-web/docroot/html/taglib/ui/input_localized/page.jsp#L21). We are noticing an issue now because we are replacing a lot of these with Clay. With this in mind, it was a reasonable approach to try to [scope React/Clay-based components with Clay-native tooltip support](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-react-web/src/main/resources/META-INF/resources/js/render.tsx#L86-L90). Unfortunately, it is not working like we would like to.

One potential solution is simply to add `lfr-portal-tooltip` class to all Clay components that have a `title`. It can be a fallback if we cannot reach a more elegant solution.

**Comment by matuzalemsteles** [comment]
Hey @bryceosterhaus instead of trying to copy some resources from `<ClayTooltipProvider />` why don't we make it flexible to allow use in `tooltip-support-web`? Since a lot of the features are similar.

> Custom renderer should be possible using ReactDOMServer.renderToString(). I have a few examples where I ported that over as well.

I remember a few discussion about this, to avoid using `ReactDOMServer.renderToString()`... but in that case maybe just using `<ClayTooltipProvider />` wouldn't it be better? I don't mean moving `<ClayTooltipProvider />` to the root of the application.

> Our current opt-in mechanism is to add lfr-portal-tooltip on element with title. Many old tags and solutions do this, see for example icon tag and input tag. We are noticing an issue now because we are replacing a lot of these with Clay.

Yeah, I think the big reason for causing a lot of problems is because we’re going to add our tooltip by default.

**Comment by bryceosterhaus** [comment]
> but in that case maybe just using <ClayTooltipProvider /> wouldn't it be better?

My primary thought on why its better to default this logic to the `tooltip-support-web` is because this touches html across the entirety of DXP and not just react generated HTML. For some reason it just _feels off_ to me to throw our DXP specific logic into the Clay component. I know thats not a great argument though lol

I think we are sort of crossing two different discussions here, we are trying to figure out both what the solution is and also where to put it. By putting it in Clay or DXP, it doesn't actually help us a ton with what the solution is.

So, taking a step back, whats the product requirement on providing tooltips? Should it work OOTB for both server rendered and js rendered markup by just adding certain html attributes? Should it be opt-in or opt-out? Not sure if @jbalsas might have a doc of requirements for this.

IMO, I think it makes for it to work globally for all `title` or `data-title` attributes and then we can provide a mechanism for customers to override this and opt-out. Tooltips for titles seem common enough that we likely would have much fewer instances of opting-out than we would for opting-in. I would think opting-out of styles is already a popular practice with css and styling in general. If we make it opt-in with `lfr-portal-tooltip` ([mentioned here](https://github.com/liferay-frontend/liferay-portal/pull/946#issuecomment-819533360)), then do we also need to back tooltip "title" attrs opt-in for our clay components?

Anyways, at the end of the day I'm not too concerned how we implement it or where it lives, I think the primary confusion is what we are trying to achieve here and what DXP's requirements for tooltips are

**Comment by markocikos** [comment]
> So, taking a step back, whats the product requirement on providing tooltips? Should it work OOTB for both server rendered and js rendered markup by just adding certain html attributes? Should it be opt-in or opt-out? Not sure if @jbalsas might have a doc of requirements for this.

@david-truong @dsanz could you please weigh in on this? See above discussion for more details.

**Comment by matuzalemsteles** [comment]
@bryceosterhaus yeah maybe I'm expressing myself badly, but really I don't want to take the DXP requirements to Clay but what I think is that we could try `TooltipProvider` to help implement `tooltip-support-web`, for example, there are some features of `TooltipProvider` that we want in the implementation of DXP, so I am imagining joining logic that makes sense in `TooltipProvider` and you can use it to implement in `tooltip-support-web`. For example, a solution that I had commented on Clay:

```jsx
<ClayTooltipProvider on={document.body}>
  {...}
</ClayTooltipProvider>
```

We are not going to change Clay's logic but just make it a little more flexible. I probably think that the errors reported by @markocikos can be corrected in Clay that are not just DXP related things but that will reflect on who uses them.

In theory, to use Tooltip we had to add `lfr-portal-tooltip`, so if we consider that we would have to keep it and on the Clay side, we can allow adding `className` to add but I'm not sure if that would be the best solution in this case. Since @bryceosterhaus investigated that `React.Portal` does not interfere with the operation of `TooltipProvider`.

**Comment by dsanz** [comment]
> > So, taking a step back, whats the product requirement on providing tooltips? Should it work OOTB for both server rendered and js rendered markup by just adding certain html attributes? Should it be opt-in or opt-out? Not sure if @jbalsas might have a doc of requirements for this.
> 
> @david-truong @dsanz could you please weigh in on this? See above discussion for more details.

I'm not aware of any document about this. If I were asked I'd lean towards an opt-in mechanism

File: modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/data_renderers/TooltipPriceRenderer.js [hunk 1]
```diff
@@ -58,17 +58,15 @@ function TooltipPriceRenderer(props) {
...
 				>
-					<span
-						className="tooltip-provider"
-						title={Liferay.Language.get('price-summary')}
-					>
-						<ClayIcon symbol="info-circle" />
-					</span>
-				</ClayTooltipProvider>
+					<ClayIcon symbol="info-circle" />
+				</span>
```
Comment (kresimir-coko) [inline modules/apps/commerce/commerce-frontend-js/src/main/resources/META-INF/resources/components/data_renderers/TooltipPriceRenderer.js:69]: How are you getting the `{Liferay.Language.get('price-summary')}` to this new tooltip? It seems that with this replacement you've lost it, or it's getting propagated to it in some other way?

File: modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/data_renderers/TooltipSummaryRenderer.js [hunk 2]
```diff
@@ -97,15 +99,9 @@ function TooltipSummaryRenderer({itemData, options, value}) {
...
 				>
-					<span
-						className="tooltip-provider"
-						title={Liferay.Language.get('info')}
-					>
-						<ClayIcon symbol="info-circle" />
-					</span>
-				</ClayTooltipProvider>
+					<ClayIcon symbol="info-circle" />
+				</span>
```
Comment (kresimir-coko) [inline modules/apps/frontend-taglib/frontend-taglib-clay/src/main/resources/META-INF/resources/data_set_display/data_renderers/TooltipSummaryRenderer.js:104]: We lose the `{Liferay.Language.get('info')}` here too, right?

**Review by kresimir-coko — COMMENTED** [review]
Hey @bryceosterhaus I left a couple comments that might be oversights. Otherwise the cutbacks on dependencies and cleaner DOM are peak performance👌

**Comment by markocikos** [comment]
> If I were asked I'd lean towards an opt-in mechanism

@bryceosterhaus I'd call that good enough to tip the scales for us, to try to go an extra mile in our solution and maintain an opt-in solution.

**Comment by bryceosterhaus** [comment]
@markocikos yep I agree, I am still trying to debug the Provider and why it's not firing for these use cases. There must be something in the React tree that is overriding and event or preventing the mouseover event to bubble up. I can't quite pinpoint it yet but I am hoping to make progress today

**Comment by bryceosterhaus** [comment]
closing in favor of https://github.com/liferay-frontend/liferay-portal/pull/1016

**Pull request closed without merging**